java/src/org/openqa/selenium/manager/SeleniumManager.java

```diff
@@ -0,0 +1,128 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.logging.Logger;
+
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+import static org.openqa.selenium.Platform.MAC;
+import static org.openqa.selenium.Platform.WINDOWS;
+
+public class SeleniumManager {
+
+    private static final Logger LOG = Logger.getLogger(SeleniumManager.class.getName());
+
+    private static final String SELENIUM_MANAGER = "selenium-manager";
+    private static final String EXE = ".exe";
+    private static final String INFO = "INFO\t";
+
+    private static SeleniumManager manager;
+
+    private File binary;
+
+    private SeleniumManager() {
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            if (binary != null && binary.exists()) {
+                try {
+                    Files.delete(binary.toPath());
+                } catch (IOException e) {
+                    LOG.warning(String.format("%s deleting temporal file: %s",
+                            e.getClass().getSimpleName(), e.getMessage()));
+                }
+            }
+        }));
+    }
+
+    public static SeleniumManager getInstance() {
+        if (manager == null) {
+            manager = new SeleniumManager();
+        }
+        return manager;
+    }
+
+    protected static String runCommand(String... command) {
+        String output = "";
+        try {
+            Process process = new ProcessBuilder(command)
+                    .redirectErrorStream(false).start();
+            process.waitFor();
+            output = CharStreams.toString(new InputStreamReader(
+                    process.getInputStream(), StandardCharsets.UTF_8));
+        } catch (InterruptedException e) {
+            LOG.warning(String.format("Interrupted exception running command %s: %s",
+                    Arrays.toString(command), e.getMessage()));
+            Thread.currentThread().interrupt();
+        } catch (Exception e) {
+            LOG.warning(String.format("%s running command %s: %s",
+                    e.getClass().getSimpleName(), Arrays.toString(command), e.getMessage()));
+        }
+        return output.trim();
+    }
+
+    protected File getBinary() {
+        if (binary == null) {
+            try {
+                Platform current = Platform.getCurrent();
+                String folder = "linux";
+                String extension = "";
+                if (current.is(WINDOWS)) {
+                    extension = EXE;
+                    folder = "windows";
+                } else if (current.is(MAC)) {
+                    folder = "mac";
+                }
+                String binaryPath = String.format("%s/%s%s", folder, SELENIUM_MANAGER, extension);
+                try (InputStream inputStream = this.getClass().getResourceAsStream(binaryPath)) {
```

Comment:
What is the advantage of copying the file before executing it?

Reply:
In Java, the binaries are distributed in a JAR file (i.e., they are compressed in these JAR). To access them, we need to access the JAR resources, getting an `InputStream` as a result. To execute this resource (a binary file), the `InputStream` should be serialized as an (executable) file. For this, it needs to be copied, and for that, a singleton pattern is the best alternative here (since we have a single instance in the JVM and we reuse the resulting binary for all the uses of the Selenium Manager singleton during all the execution, deleting the temporal file on shutdown).

Reply:
Completely agree with @bonigarcia, makes sense. I would only ask to write a brief version of this comment as comment(s) in the code for future reference.

Reply:
Done